### DISCUSSION TITLE
Finite Viscoelastic behaviors

### DISCUSSION THREAD

**Discussion by bhaveshshrimali** [thelfer/tfel#154] [Q&A]
Migrated from [here](https://github.com/thelfer/MFrontGenericInterfaceSupport/issues/94#event-6257260183). 

I am interested in exploring `Mfront` to implement finite-viscoelastic behaviors. The idea is to implement the model in [this paper](http://pamies.cee.illinois.edu/Publications_files/CRM_2016.pdf) specifically equations (37) and (41) for the internal variable using `Mfront/mgis`. I implemented the explicit integration in dolfinx, however the resulting code is humongous and the performance is highly variable. 

What would you recommend as a good starting point to start implementing the model above? I am looking at the isotropic linear harderning plasticity mfront code, namely

```cpp
@DSL DefaultDSL;
@Behaviour IsotropicLinearHardeningPlasticity;
@Author Thomas Helfer;
@Date   14/10/2016;

@Description{
  An implicit implementation of a simple
  isotropic plasticity behaviour with
  isotropic linear hardening.

  The yield surface is defined by:
  "\["
  "  f(\sigmaeq,p) = \sigmaeq-s_{0}-H\,p"
  "\]"
}

@MaterialProperty stress young;
young.setGlossaryName("YoungModulus");
@MaterialProperty real nu;
nu.setGlossaryName("PoissonRatio");
@MaterialProperty stress H;
H.setEntryName("HardeningSlope");
@MaterialProperty stress s0;
s0.setGlossaryName("YieldStress");

@StateVariable StrainStensor eel;
eel.setGlossaryName("ElasticStrain");
@StateVariable strain p;
p.setGlossaryName("EquivalentPlasticStrain");

/*!
 * computation of the prediction operator: we only provide the elastic
 * operator.
 *
 * We could also provide a tangent operator, but this would mean
 * saving an auxiliary state variable stating if a plastic loading
 * occured at the previous time step.
 */
@PredictionOperator{
  // silent "unused parameter" warning
  static_cast<void>(smt);
  const auto lambda = computeLambda(young,nu);
  const auto mu     = computeMu(young,nu);
  Dt = lambda*Stensor4::IxI()+2*mu*Stensor4::Id();
}

/*!
 * behaviour integration using a fully implicit Euler-backwark scheme.
 */
@ProvidesSymmetricTangentOperator;
@Integrator{
  const auto lambda = computeLambda(young,nu);
  const auto mu     = computeMu(young,nu);
  eel += deto;
  const auto se     = 2*mu*deviator(eel);
  const auto seq_e  = sigmaeq(se);
  const auto b      = seq_e-s0-H*p>stress{0};
  if(b){
    const auto iseq_e = 1/seq_e;
    const auto n      = eval(3*se/(2*seq_e));
    const auto cste   = 1/(H+3*mu);
    dp   = (seq_e-s0-H*p)*cste;
    eel -= dp*n;
    if(computeTangentOperator_){
      if(smt==CONSISTENTTANGENTOPERATOR){
	Dt = (lambda*Stensor4::IxI()+2*mu*Stensor4::Id()
	      -4*mu*mu*(dp*iseq_e*(Stensor4::M()-(n^n))+cste*(n^n)));
      } else {
	Dt = lambda*Stensor4::IxI()+2*mu*Stensor4::Id();
      }
    }
  } else {
    if(computeTangentOperator_){
      Dt = lambda*Stensor4::IxI()+2*mu*Stensor4::Id();
    }
  }
  sig = lambda*trace(eel)*Stensor::Id()+2*mu*eel;
}

```

and the tests https://github.com/thelfer/MFrontGenericInterfaceSupport/tree/master/tests

A couple of high-level points: 

  * I do not explicitly need a strain-measure per-se. My deformation gradient, which can be multiplicatively decomposed as F = Fe  * Fv (as Thomas correrctly pointed out) is sufficient for my analysis. 
  
   * For the evolution law (and constitutive relation) the internal variable is the symmetric part of Fv, namely Cv (= Fv^T * Fv). So in short I have the constitutive relation as well as the evolution equation in terms of Cv.

I hope this is sufficient to describe what I want to achieve. As an aside, the specific constitutive model is [here](https://github.com/bhaveshshrimali/FEniCS_Shrimali_Ghosh_Kumar_Lopez-Pamies/blob/main/src/viscoFEM_dolfinx.py) in case if someone wants to take a look

Thanks

**Comment by thelfer** [comment]
@bhaveshshrimali Thanks for your interest for `MFront` and `MGIS`.

First of all, I must warn you that we did not yet port `mgis.fenics` to `dolfin-x`. @bleyrj and myself plan to do this but we have not yet found the time to do so. Any help would be kindly appreciated. I do believe that you can more or less adapt Jérémy's tutorial: <https://comet-fenics.readthedocs.io/en/latest/demo/plasticity_mfront/plasticity_mfront.py.html>

For the starting point, you can have a look at the MFront gallery for hyperelastic behaviours (<https://thelfer.github.io/tfel/web/gallery.html#hyperelasticity>) and hyperviscoelastic behaviours (<https://thelfer.github.io/tfel/web/hyperviscoelasticity.html>).

I am currently writting a tutorial about FeFp implementations in `MFront` (plasticity, viscoplasticity and single crystal plasticity). I can attach my draft if you want.

**Reply by bhaveshshrimali** [reply]
> First of all, I must warn you that we did not yet port mgis.fenics to dolfin-x. @bleyrj and myself plan to do this but we have not yet found the time to do so. Any help would be kindly appreciated.

Sure, once I familiarize myself with mfront and the associated code-generation process, I can maybe take a stab at extending it to dolfinx. My understanding of inner dolfinx workings isn't super-complete but I am familiar more-or-less with the python bindings at least, and with some help I should be able to navigate my way around it. 

> I do believe that you can more or less adapt Jérémy's tutorial: [comet-fenics.readthedocs.io/en/latest/demo/plasticity_mfront/plasticity_mfront.py.html](https://comet-fenics.readthedocs.io/en/latest/demo/plasticity_mfront/plasticity_mfront.py.html)

Thanks, I'll take a detailed look. I have taken a look at this before but I don't remember what was missing back then that I didn't use it.

> For the starting point, you can have a look at the MFront gallery for hyperelastic behaviours ([thelfer.github.io/tfel/web/gallery.html#hyperelasticity](https://thelfer.github.io/tfel/web/gallery.html#hyperelasticity)) and hyperviscoelastic behaviours ([thelfer.github.io/tfel/web/hyperviscoelasticity.html](https://thelfer.github.io/tfel/web/hyperviscoelasticity.html)).

Sure, thanks! From the look of it the viscoelastic behavior is more along the lines of hereditary integrals and not on internal variables (like plasticity). But it should help me familiarize myself nonetheless. 

  > I am currently writting a tutorial about FeFp implementations in MFront (plasticity, viscoplasticity and single crystal plasticity). I can attach my draft if you want.
  
 That would be a great help! :) Thanks

**Reply by thelfer** [reply]
Please find attached a draft, for what it's worth. If your behaviour fits into this framework, may be we could add a section about it

[FeFp.zip](https://github.com/thelfer/tfel/files/8288471/FeFp.zip)
